crud_int_test.go

```diff
@@ -0,0 +1,95 @@
+package gocbcorex_test
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/couchbase/gocbcorex"
+	"github.com/couchbase/gocbcorex/testutilsint"
+	"github.com/google/uuid"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+const NUM_REPLICAS = 2
+
+func TestGetAllReplicasDino(t *testing.T) {
+	opts := CreateDefaultAgentOptions()
+	agent, err := gocbcorex.CreateAgent(context.Background(), opts)
+	defer func() {
+		err := agent.Close()
+		require.NoError(t, err)
+	}()
+
+	require.NoError(t, err)
+
+	docKey := uuid.NewString()
+	upsertRes, err := agent.Upsert(context.Background(), &gocbcorex.UpsertOptions{
+		Key:            []byte(docKey),
+		ScopeName:      "",
+		CollectionName: "",
+		Value:          []byte(`{"foo": "bar"}`),
+	})
+	require.NoError(t, err)
+	assert.NotZero(t, upsertRes.Cas)
+
+	time.Sleep(time.Second * 2)
+
+	ctx := context.Background()
+	stream, err := agent.GetAllReplicas(ctx, &gocbcorex.GetAllReplicasOptions{
+		Key:            []byte(docKey),
+		BucketName:     "default",
+		ScopeName:      "",
+		CollectionName: "",
+	})
+	require.NoError(t, err)
+
+	result, err := stream.Next()
+	masterCount := 0
+	replicaCount := 0
+	for result != nil {
+		require.NoError(t, err)
+		require.Equal(t, result.Value, []byte(`{"foo": "bar"}`))
+		if !result.IsReplica {
+			masterCount += 1
+		} else {
+			replicaCount += 1
+		}
+		result, err = stream.Next()
+	}
+	require.Equal(t, NUM_REPLICAS, masterCount+replicaCount)
+
+	testutilsint.SkipIfNoDinoCluster(t)
+	nodes := testutilsint.GetTestNodes(t)
+	nonOrchestrators := nodes.Select(func(node *testutilsint.NodeTarget) bool {
+		return !node.IsOrchestrator
+	})
+
+	dino := testutilsint.StartDinoTesting(t, true)
+	for _, node := range nonOrchestrators {
+		dino.BlockAllTraffic(node.Hostname)
```

Comment:
In this line, you're blocking all non-orchestrators from talking to eachother, but you've already previously run the Upsert and then the GetAllReplicas.  This means it's likely that the document will already have propagated to all the nodes, but more importantly it means that the following operation should be guarenteed to fail since it will be unable to contact the majority of the nodes.  A more representative test might be to block one node that is a replica for a document, perform an upsert to the primary node, and then GetAllReplicas.  We would expect that the GetAllReplicas operation would fail until you've unblocked the final node from the replicas.

Reply:
With this I wanted the replica to already be present on these nodes, but I wanted them to be unreachable when we run the GetAllReplicas. I wouldn't expect the command to fail I'd just expect the single document from the orchestrator to be returned. I wasn't aware that we wanted the get all Replicas to fail on unreachable nodes, I thought the plan was to return all available replicas, allowing the user to continue in the case where they are using this to get any available replica.